cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/DefaultStore.java

```diff
@@ -559,6 +559,29 @@ public void scanApplications(int txBatchSize, BiConsumer<ApplicationId, Applicat
     }
   }
 
+  @Override
+  public void scanApplications(NamespaceId id, int txBatchSize,
+                               BiConsumer<ApplicationId, ApplicationSpecification> consumer) {
+
+    AtomicReference<AppMetadataStore.Cursor> cursorRef = new AtomicReference<>(AppMetadataStore.Cursor.EMPTY);
+
+    while (true) {
+      AtomicInteger count = new AtomicInteger();
+
+      TransactionRunners.run(transactionRunner, context -> {
+        getAppMetadataStore(context).scanApplications(id.getNamespace(), cursorRef.get(), (cursor, entry) -> {
```

Comment:
Seems like the only diff is calling the AppMetadataStore.scanApplications with or without the namespaceId? Maybe we can just have one scanApplications method in the AppMetadataStore that take namespaceId optionally?

Reply:
By optionally, do you mean Optional<String> NamespaceId as an argument for scanApplications?